tests_e2e/playwright/basic.py

```diff
@@ -0,0 +1,288 @@
+        self.page.get_by_text("File", exact=True).click()
+        # close all tabs
+        self.page.get_by_role("menuitem", name="Close All Tabs", exact=True).click()
+
+        # there may be a popup to save your work, don't save
+        if self.page.get_by_text("Save your work", exact=True).is_visible():
+            self.page.get_by_role("button", name="Discard", exact=True).click()
+
+    def _set_environment_via_popup(self, kernel=None):
+        """Set the environment kernel via the popup dialog box.
+        If kernel is `None`, `No Kernel` is selected and the popup is dismissed.
+
+        Attributes
+        ----------
+        kernel: str or None
+            (Optional) name of conda environment to set. Defaults to None.
+
+        """
+        if kernel is None:
+            # close dialog (deal with the two formats of this dialog)
+            try:
+                self.page.get_by_text("Cancel", exact=True).click()
+            except:
+                self.page.locator("div").filter(has_text="No KernelSelect").get_by_role(
+                    "button", name="No Kernel"
+                ).wait_for(timeout=300, state="attached")
+        else:
+            # set the environment
+            self.page.get_by_role("combobox").nth(1).select_option(
+                f'{{"name":"{kernel}"}}'
+            )
+            # click Select to close popup (deal with the two formats of this dialog)
+            try:
+                self.page.get_by_role("button", name="Select", exact=True).click()
+            except:
+                self.page.locator("div").filter(has_text="No KernelSelect").get_by_role(
+                    "button", name="Select"
+                ).click()
+
+    def set_environment(self, kernel):
+
+        popup = self._check_for_kernel_popup()
+        # if there is not a kernel popup, make it appear
+        if not popup:
+            self.page.get_by_text("Kernel", exact=True).click()
+            self.page.get_by_role("menuitem", name="Change Kernel…").get_by_text(
+                "Change Kernel…"
+            ).click()
+
+        self._set_environment_via_popup(kernel)
+
+
+class RunNotebook(Navigator):
+    def run_notebook(self, path, expected_output_text):
+        logger.debug(">>> Run test notebook")
+
+        # TODO: add nbgitpuller here?
+
+        # navigate to specific notebook
+        file_locator = self.page.get_by_text("File", exact=True)
+
+        file_locator.wait_for(
+            timeout=300000, state="attached"
+        )  # 5 minutes max for server to spin up
+        file_locator.click()
+        self.page.get_by_role("menuitem", name="Open from Path…").get_by_text(
+            "Open from Path…"
+        ).click()
+        self.page.get_by_placeholder("/path/relative/to/jlab/root").fill(path)
+        self.page.get_by_role("button", name="Open").click()
+        # give the page a second to open, otherwise the options in the kernel menu will be disabled.
+        time.sleep(1)
```

Comment:
This is fragile. I would rely on the the locator api to check that a locator is e.g. visible. https://playwright.dev/docs/api/class-locator#locator-is-visible

Reply:
I agree its fragile. I struggled getting `is_visible` and `wait_for` to work with pages loading, particularly after redirect. I will go back and give it more effort to see if I can get rid of the sleeps.

Reply:
I resolved this using `self.page.wait_for_load_state('networkidle')`.